src/Npgsql/NpgsqlRecordDataReader.cs

```diff
@@ -0,0 +1,327 @@
+﻿using Npgsql.Internal;
+using Npgsql.Internal.TypeHandling;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeMapping;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Npgsql
+{
+    internal sealed class NpgsqlRecordDataReader : DbDataReader
+    {
+        NpgsqlDataReader _outermostReader;
+        ulong _uniqueOutermostReaderRowId;
+        int _depth;
+        int _numRows;
+        int _nextRowIndex;
+        int _nextRowBufferPos;
+
+        ColumnInfo[]? _columns;
+
+        bool _calledNextResult;
+        bool _closed;
+
+        struct ColumnInfo
+        {
+            public uint Oid;
+            public int BufferPos;
+        }
+
+        NpgsqlReadBuffer Buffer => _outermostReader.Buffer;
+        ConnectorTypeMapper TypeMapper => _outermostReader.Connector.TypeMapper;
+
+        internal NpgsqlRecordDataReader(NpgsqlDataReader outermostReader, ulong uniqueOutermostReaderRowId, int depth)
+        {
+            _outermostReader = outermostReader;
+            _uniqueOutermostReaderRowId = uniqueOutermostReaderRowId;
+            _depth = depth;
+        }
+
+        internal void InitArray()
+        {
+            var dimensions = Buffer.ReadInt32();
+            var containsNulls = Buffer.ReadInt32() == 1;
+            Buffer.ReadUInt32(); // Element OID. Ignored.
+
+            if (containsNulls)
+                throw new InvalidOperationException("Record array contains null record");
+
+            if (dimensions == 0)
+                return;
+
+            if (dimensions != 1)
+                throw new InvalidOperationException($"Cannot read an array with 1 dimension from an array with {dimensions} dimension(s)");
+
+            _numRows = Buffer.ReadInt32();
+            Buffer.ReadInt32(); // Lower bound
+
+            if (_numRows > 0)
+                Buffer.ReadInt32(); // Length of first row
+
+            _nextRowBufferPos = Buffer.ReadPosition;
+        }
+
+        internal void InitSingleRow()
+        {
+            _numRows = 1;
+            _nextRowBufferPos = Buffer.ReadPosition;
+        }
+
+        public override object this[int ordinal] => GetValue(ordinal);
+
+        public override object this[string name] => GetValue(GetOrdinal(name));
+
+        public override int Depth
+        {
+            get
+            {
+                CheckNotClosed();
+                return _depth;
+            }
+        }
+
+        public override int FieldCount
+        {
+            get
+            {
+                CheckNotClosed();
+                return _calledNextResult ? 0 : _columns?.Length ?? 0;
```

Comment:
The behavior in both Npgsql and SqlClient is for FieldCount to return the correct field count before the first Read is called; this is because the resultset obviously has the same field count for all rows.

Here the situation is different - rows can actually have different field counts (heterogeneous rows), so I think this makes sense.

Note: we may want to add a test for heterogeneous rows and make sure this is properly supported, with FieldCount returning the right thing.